cxplat/src/cxplat_winuser/workitem_winuser.cpp

```diff
@@ -0,0 +1,164 @@
+    UNREFERENCED_PARAMETER(work);
+
+    if (parameter == nullptr) {
+        return;
+    }
+
+    cxplat_preemptible_work_item_t* work_item = (cxplat_preemptible_work_item_t*)parameter;
+    work_item->work_item_routine(work_item->work_item_context);
+
+    cxplat_free_preemptible_work_item(work_item);
+}
+
+_Must_inspect_result_ cxplat_status_t
+cxplat_allocate_preemptible_work_item(
+    _In_opt_ void* caller_context,
+    _Outptr_ cxplat_preemptible_work_item_t** work_item,
+    _In_ void (*work_item_routine)(_Inout_opt_ void* work_item_context),
+    _Inout_opt_ void* work_item_context)
+{
+    UNREFERENCED_PARAMETER(caller_context);
+    cxplat_status_t result = CXPLAT_STATUS_SUCCESS;
+
+    if (!cxplat_acquire_rundown_protection(&_cxplat_preemptible_work_items_rundown_reference)) {
+        return CXPLAT_STATUS_UNSUCCESSFUL;
+    }
+
+    *work_item = (cxplat_preemptible_work_item_t*)cxplat_allocate(sizeof(cxplat_preemptible_work_item_t));
+    if (*work_item == nullptr) {
+        result = CXPLAT_STATUS_NO_MEMORY;
+        goto Done;
+    }
+
+    // It is required to use the InitializeThreadpoolEnvironment function to
+    // initialize the _callback_environment structure before calling CreateThreadpoolWork.
+    (*work_item)->work = CreateThreadpoolWork(_cxplat_preemptible_routine, *work_item, &_callback_environment);
+    if ((*work_item)->work == nullptr) {
+        result = CXPLAT_STATUS_FROM_WIN32(GetLastError());
+        goto Done;
+    }
+    (*work_item)->work_item_routine = work_item_routine;
+    (*work_item)->work_item_context = work_item_context;
+
+Done:
+    if (result != CXPLAT_STATUS_SUCCESS) {
+        cxplat_release_rundown_protection(&_cxplat_preemptible_work_items_rundown_reference);
+        cxplat_free(*work_item);
+        *work_item = nullptr;
+    }
+    return result;
+}
+
+void
+cxplat_queue_preemptible_work_item(_Inout_ cxplat_preemptible_work_item_t* io_work_item)
+{
+    SubmitThreadpoolWork(io_work_item->work);
+}
+
+void
+cxplat_free_preemptible_work_item(_Frees_ptr_opt_ cxplat_preemptible_work_item_t* work_item)
+{
+    if (!work_item) {
+        return;
+    }
+
+    CloseThreadpoolWork(work_item->work);
+    cxplat_free(work_item);
+
+    cxplat_release_rundown_protection(&_cxplat_preemptible_work_items_rundown_reference);
+}
+
+cxplat_status_t
+cxplat_winuser_initialize_thread_pool()
+{
+    cxplat_status_t status = CXPLAT_STATUS_SUCCESS;
+    bool cleanup_group_created = false;
+    bool return_value;
+
+    cxplat_initialize_rundown_protection(&_cxplat_preemptible_work_items_rundown_reference);
+
+    // Initialize a callback environment for the thread pool.
+    InitializeThreadpoolEnvironment(&_callback_environment);
+
+    _pool = CreateThreadpool(nullptr);
+    if (_pool == nullptr) {
+        status = CXPLAT_STATUS_FROM_WIN32(GetLastError());
+        goto Exit;
+    }
+
+    SetThreadpoolThreadMaximum(_pool, 1);
+    return_value = SetThreadpoolThreadMinimum(_pool, 1);
```

Comment:
Perhaps some code comments to explain why this is done?

Reply:
@Alan-Jowett was this originally your code?

Reply:
The minimum was to ensure that we make forward progress on low memory. The maximum was to catch cases where work-items depend on each other.

I.e. if work item A blocks while executing waiting on work item B, then it may deadlock on low memory. Better to deadlock all the time.

Reply:
Sounds like another case of the code being debug-focused, which is fine for debug builds, but not retail/release builds.

Reply:
sure: maybe Alan can add some `ifndef NDEBUG` around such code?